cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -324,25 +324,146 @@ module ProductFlow {
 
+    private predicate isSourcePair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    private predicate isSinkPair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSinkPair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    pragma[assume_small_delta]
     pragma[nomagic]
-    private predicate reachableInterprocEntry(
-      Flow1::PathNode source1, Flow2::PathNode source2, Flow1::PathNode node1, Flow2::PathNode node2
-    ) {
-      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState()) and
-      node1 = source1 and
-      node2 = source2
+    private predicate fwdReachableInterprocEntry(Flow1::PathNode node1, Flow2::PathNode node2) {
+      isSourcePair(node1, node2)
+      or
+      fwdIsSuccessor(_, _, node1, node2)
+    }
+
+    pragma[assume_small_delta]
+    pragma[nomagic]
+    private predicate fwdLocalPathStep1(Flow1::PathNode n) {
+      fwdReachableInterprocEntry(n, _)
+      or
+      exists(Flow1::PathNode mid |
+        fwdLocalPathStep1(mid) and
+        localPathStep1(mid, n)
+      )
+    }
+
+    pragma[assume_small_delta]
+    pragma[nomagic]
+    private predicate revLocalPathStep1(Flow1::PathNode n) {
+      fwdLocalPathStep1(n) and
+      (
+        isSinkPair(n, _)
+        or
+        interprocEdgePair(n, _, _, _)
+        or
+        exists(Flow1::PathNode mid |
+          revLocalPathStep1(mid) and
+          localPathStep1(n, mid)
+        )
+      )
+    }
+
+    pragma[assume_small_delta]
+    private predicate prunedLocalPathStep1(Flow1::PathNode n1, Flow1::PathNode n2) {
+      revLocalPathStep1(n1) and
+      revLocalPathStep1(n2) and
+      localPathStep1(n1, n2)
+    }
+
+    pragma[nomagic]
+    private predicate fwdLocalPathStep2(Flow2::PathNode n) {
+      fwdReachableInterprocEntry(_, n)
       or
-      exists(
-        Flow1::PathNode midEntry1, Flow2::PathNode midEntry2, Flow1::PathNode midExit1,
-        Flow2::PathNode midExit2
+      exists(Flow2::PathNode mid |
+        fwdLocalPathStep2(mid) and
+        localPathStep2(mid, n)
+      )
+    }
+
+    pragma[assume_small_delta]
+    pragma[nomagic]
+    private predicate revLocalPathStep2(Flow2::PathNode n) {
+      fwdLocalPathStep2(n) and
+      (
+        isSinkPair(_, n)
+        or
+        interprocEdgePair(_, n, _, _)
+        or
+        exists(Flow2::PathNode mid |
+          revLocalPathStep2(mid) and
+          localPathStep2(n, mid)
+        )
+      )
+    }
+
+    pragma[assume_small_delta]
+    private predicate prunedLocalPathStep2(Flow2::PathNode n1, Flow2::PathNode n2) {
+      revLocalPathStep2(n1) and
+      revLocalPathStep2(n2) and
+      localPathStep2(n1, n2)
+    }
+
+    private predicate localPathStep1SuccPlus(Flow1::PathNode n1, Flow1::PathNode n2) =
+      fastTC(prunedLocalPathStep1/2)(n1, n2)
```

Comment:
Wait a minute, are you doing recursion through fastTC? That's going to be slow.

Reply:
https://github.com/github/codeql-core/issues/3098

Reply:
Ah. TIL! Is it safe to replace the `fastTC` with a transitive closure operator? Or should I write out the TC manually to prevent the compiler from generating a `fastTC`?

Reply:
> Is it safe to replace the `fastTC` with a transitive closure operator?

If that's the only change, then no, that just puts the choice in the compilers hands about whether to use fastTC or unrolling. Manually writing out the TC might also not be the best, but that's hard to tell. At least it wouldn't be the first option I'd try. The main issue here is a proliferation of recursive dependencies in the pruning of your TCs. If you overapproximate these slightly then you may are able to do source-sink-restricted materialisations of the TCs that are independent of the recursion. Alternatively, the "speed up isSuccessor" commit that you've felt compelled to add, might not be needed at all if you tweak the join-order - currently the join-order is likely to be somewhat restricted, since the original two applications of TC are unlikely to distribute over the disjunction, which will invariably lead to a poor join-order.

Reply:
Yeah, the motivation for that commit was that I was seeing some very large blowups on a particular project:
```ql
Tuple counts for ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessor#4#ffff/4@i10#848b8wcn after 12m10s:
  2500       ~0%        {4} r1 = SCAN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta OUTPUT In.0 'pred1', In.1 'pred2', In.0 'pred1', In.1 'pred2'
  
  1498000    ~0%        {4} r2 = JOIN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep1#2Plus#ff ON FIRST 1 OUTPUT Rhs.1 'succ1', Lhs.1 'pred2', Lhs.0 'pred1', Lhs.1 'pred2'
  
  2500       ~3%        {3} r3 = SCAN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta OUTPUT In.1 'pred2', In.0 'pred1', In.0 'pred1'
  
  1211000    ~0%        {3} r4 = JOIN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep1#2Plus#ff ON FIRST 1 OUTPUT Lhs.1 'pred2', Lhs.0 'pred1', Rhs.1 'succ1'
  
  1213500    ~0%        {3} r5 = r3 UNION r4
  
  3829668500 ~2%        {4} r6 = JOIN r5 WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep2#2Plus#ff ON FIRST 1 OUTPUT Lhs.2 'succ1', Rhs.1 'succ2', Lhs.1 'pred1', Lhs.0 'pred2'
  
  3831166500 ~2%        {4} r7 = r2 UNION r6
  3831169000 ~2%        {4} r8 = r1 UNION r7
  0          ~0%        {4} r9 = JOIN r8 WITH ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::isSinkPair#2#ff ON FIRST 2 OUTPUT Lhs.2 'pred1', Lhs.3 'pred2', Lhs.0 'succ1', Lhs.1 'succ2'
  
  2000       ~0%        {4} r10 = SCAN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta OUTPUT In.0 'pred1', In.0 'pred1', In.1 'pred2', In.1 'pred2'
  
  69000      ~0%        {4} r11 = JOIN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep1#2Plus#ff ON FIRST 1 OUTPUT Rhs.1 'succ1', Lhs.0 'pred1', Lhs.1 'pred2', Lhs.1 'pred2'
  
  3828324000 ~3%        {4} r12 = JOIN r5 WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep2#2Plus#ff ON FIRST 1 OUTPUT Lhs.2 'succ1', Lhs.1 'pred1', Lhs.0 'pred2', Rhs.1 'succ2'
  
  3828393000 ~3%        {4} r13 = r11 UNION r12
  3828395000 ~3%        {4} r14 = r10 UNION r13
  765712148  ~2%        {7} r15 = JOIN r14 WITH ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::interprocEdge1#5#fffff_20134#join_rhs ON FIRST 1 OUTPUT Rhs.1, Rhs.2, Lhs.3, Rhs.4, Lhs.1 'pred1', Lhs.2 'pred2', Rhs.3 'succ1'
  241499     ~9311%     {4} r16 = JOIN r15 WITH ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::interprocEdge2#5#fffff_01243#join_rhs ON FIRST 4 OUTPUT Lhs.4 'pred1', Lhs.5 'pred2', Lhs.6 'succ1', Rhs.4 'succ2'
  
  241499     ~9311%     {4} r17 = r9 UNION r16
  3016       ~18%       {4} r18 = r17 AND NOT ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessor#4#ffff#prev(Lhs.0 'pred1', Lhs.1 'pred2', Lhs.2 'succ1', Lhs.3 'succ2')
                        return r18
```
and it was _slightly_ mitigated 594da1a21aa7a6ae271d23ba25724b8ab3399bd4. However, the following join order tweaks removes that problem altogether, and I don't even need that commit anymore 🎉
```ql
Pipeline standard for ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessor#4#ffff@56f5ewp5 was evaluated in 7 iterations totaling 161ms (delta sizes total: 95745).
                    {4} r1 = ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessorExit#4#ffff UNION _#ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPoi__#join_rhs
                
  57775     ~61%    {4} r2 = JOIN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta WITH r1 ON FIRST 2 OUTPUT Lhs.0, Lhs.1, Rhs.2, Rhs.3
  57775     ~61%    {4} r3 = r2 AND NOT ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessor#4#ffff#prev(Lhs.0, Lhs.1, Lhs.2, Lhs.3)
                
  1058085      ~0%    {5} r4 = JOIN ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdReachableInterprocEntry#2#ff#prev_delta WITH _#ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPoi__#join_rhs#1 ON FIRST 1 OUTPUT Lhs.1, Rhs.3, Rhs.1, Rhs.2, Lhs.0
  1058085      ~0%    {5} r5 = r4 AND NOT ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::fwdIsSuccessor#4#ffff#prev(Lhs.4, Lhs.0, Lhs.2, Lhs.3)
  665971  ~12936%    {4} r6 = JOIN r5 WITH #ProductFlow#20473ca4::ProductFlow::GlobalWithState#InvalidPointerDeref#e13af12b::AllocToInvalidPointerConfig#::localPathStep2#2Plus#ff ON FIRST 2 OUTPUT Lhs.4, Lhs.0, Lhs.2, Lhs.3
                
  723746  ~11576%    {4} r7 = r3 UNION r6
                    return r7
```